### PR TITLE
Enable Spark Job Remote Debug in Executor context

### PR DESCRIPTION
**Pull request by wezhang**
The Spark Remote Debugger Runner is refactory for support
multiple debug processes.

Signed-off-by: Zhang Wei <zhwe@microsoft.com>

### CODE REVIEW

**Comment by wezhang** [comment]
[ERROR] Internal error: java.lang.RuntimeException: Failed to load p2 repository with ID 'mars' from location http://download.eclipse.org/releases/mars: Unable to read repository at http://download.eclipse.org/releases/mars/content.xml. Connect to download.eclipse.org:80 timed out -

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/common/HDInsightUtil.java [hunk 1]
```diff
@@ -165,8 +165,10 @@ public static void showWarningMessageOnSubmissionMessageWindow(@NotNull final Pr
     private static void showInfoOnSubmissionMessageWindow(@NotNull final Project project, @NotNull final MessageInfoType type, @NotNull final String message, @NotNull final boolean isNeedClear) {
 
         synchronized (LOCK) {
-            final ToolWindow toolWindow = ToolWindowManager.getInstance(project).getToolWindow(CommonConst.SPARK_SUBMISSION_WINDOW_ID);
-            showSubmissionMessage(toolWindow, project, message, type, isNeedClear);
+            try {
+                final ToolWindow toolWindow = ToolWindowManager.getInstance(project).getToolWindow(CommonConst.SPARK_SUBMISSION_WINDOW_ID);
+                showSubmissionMessage(toolWindow, project, message, type, isNeedClear);
+            } catch (Exception ignore) { /* IDE disposed */ }
```
Comment (voidfunction) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/common/HDInsightUtil.java]: Could you help comment some details here?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/SparkBatchJobDebuggerRunner.java [hunk 2]
```diff
@@ -50,33 +54,56 @@
...
-    private PublishSubject<DebugAction> actionSubject = PublishSubject.create();
+    private SparkBatchDebugSession debugSession;
+    private boolean isAppInsightEnabled = true;
+    private Phaser debugProcessPhaser;
+    private SparkBatchRemoteDebugJob debugJob;
 
     // More complex pattern, please use grok
-    private Pattern simpleLogPattern = Pattern.compile("\\d{1,2}[/-]\\d{1,2}[/-]\\d{1,2} \\d{1,2}:\\d{1,2}:\\d{1,2} (INFO|WARN|ERROR) .*", Pattern.DOTALL);
+    private final Pattern simpleLogPattern = Pattern.compile("\\d{1,2}[/-]\\d{1,2}[/-]\\d{1,2} \\d{1,2}:\\d{1,2}:\\d{1,2} (INFO|WARN|ERROR) .*", Pattern.DOTALL);
+    private final Pattern executorLogUrlPattern = Pattern.compile("^\\s+SPARK_LOG_URL_STDERR -> https?://([^:]+):?\\d*/node/containerlogs/(container.*)/livy/stderr.*");
```
Comment (voidfunction) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/SparkBatchJobDebuggerRunner.java]: static?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/SparkBatchJobDebuggerRunner.java [hunk 3]
```diff
@@ -96,91 +123,71 @@ public GenericDebuggerRunnerSettings createConfigurationData(ConfigurationInfoPr
...
-        Map<String, String> postEventProperty = new HashMap<>();
+        Project project = submitModel.getProject();
+        JobStatusManager jobStatusMgmt = HDInsightUtil.getSparkSubmissionToolWindowManager(project)
+                .getJobStatusManager();
+
+        // Reset the debug process Phaser
+        debugProcessPhaser = new Phaser(1);
+
+        Observable.create((Observable.OnSubscribe<String>) ob ->
+                createDebugJobSession(submitModel).subscribe(debugJobClusterPair-> {
```
Comment (voidfunction) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/SparkBatchJobDebuggerRunner.java]: >ir- [](start = 78, length = 3)

space?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/common/HDInsightUtil.java (hunk 1 above)
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/common/HDInsightUtil.java]: I hit the issue once closing the Intellij, there was a AssertionError "Already disposed" thrown from `ToolWindowManager.getInstance(project)`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/SparkBatchJobDebuggerRunner.java (hunk 3 above)
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/SparkBatchJobDebuggerRunner.java]: Will fix it in next change.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/SparkBatchJobDebuggerRunner.java [hunk 4]
```diff
@@ -50,33 +54,56 @@
...
-import java.util.Optional;
+import java.util.*;
+import java.util.concurrent.Phaser;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
+import static rx.exceptions.Exceptions.propagate;
+
 public class SparkBatchJobDebuggerRunner extends GenericDebuggerRunner {
```
Comment (djlan) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/SparkBatchJobDebuggerRunner.java]: >SparkBatchJobDebuggerRunner [](start = 13, length = 27)

Should we add a option (like checkbox) in submission UI? Debug executor is only enabled after it is checked and by default set the number of executor to 1?

Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/SparkBatchJobDebuggerRunner.java]: Currently, no limitation to executor number. I've tried running with default value, so far so good.

**Pull request merged**